### ISSUE TITLE
Some race conditions

### ISSUE THREAD

**Issue by jamonkko** [tdtran/resque-uniq#10]
There are 2 possible race conditions currently:
a) If lock is expired manually (in get_lock) by checking the ttl there is no guarantee that no other process has modified (e.g. deleted and set) between getting the value and deleting it. So a new lock could be accidentally deleted.
b) If two or more processes are same time enqueuing and checking stale locks and they both notice that lock is stale (stale_lock? returns true for both of them because they are running it at the same time). And after that the first one is faster to deleting the locks and sets a new value for lock. And then the second one goes forward and deletes the new one set by first process and sets again a new value. Then both jobs will be in queue and if there are more than one worker for the queue the jobs are performed parallel.

We suspect that we saw the b) happening in our system. Of course it is hard to reproduce or prove :) But looking at the code it seems that in these quite rare circumstances resque-uniq does not work correctly at the moment.

I created a branch that fixes this problem by running those operations by evaluating atomic LUA scripts in redis. Problem is that it is supported only starting from Redis 2.6, so this change would mean that resque-uniq would not work with older Redis versions.

I will test that in practice still before creating pull request.

**Comment by tume** [comment]
Any progress on the pull request? We are also experiencing this in our production system.

**Comment by tdtran** [comment]
I haven't investigated this issue enough to say for sure. We haven't run into it in our production.

@jamonkko you said you created a branch to fix it. Using Lua script is probably a bit heavy but I would like to see your patch anyway. Can you submit a pull request and attach that Lua script? Thanks

**Comment by jamonkko** [comment]
Ok, I will submit it. I was hoping to run it in real environment for some time before submitting to make sure it doesn't have any problems. But I guess it is been long enough.
I also added there enough unit tests to make sure Lua scripting parts have good test coverage.

I did consider using just Redis api to fix it, maybe with GETSET command or with MULTI transaction, but using either one seemed to make the implementation even more complex.
So using Lua script to do atomic operations was simpler, although little bit heavy.
